### ISSUE TITLE
Exception when saving product with tier price on website scope

### ISSUE THREAD

**Issue by johan-lindahl** [magento/magento2#35444]
### Preconditions (*)

Magento 2.4.3-p2

The problem occurs when the store has "price scope" = website and the product have tier prices on multiple websites.

### Steps to reproduce (*)

1.  Read a product and explicitly specify the default scope
$product = $this->productRepository->get($MySku, true, \Magento\Store\Model\Store::DEFAULT_STORE_ID);
(productRepository = \Magento\Catalog\Model\ProductRepository)

2. Set some attribute e.g. $product->setName($name);

3. Save the product
$this->productRepository->save($product);

### Expected result (*)

1. The product should be saved and no exception should be thrown.

### Actual result (*)

1. The following exception is thrown,

"SQLSTATE[23000]: Integrity constraint violation: 1062 Duplicate entry '1021-0-274-1.0000-4' for key 'UNQ_EBC6A54F44DFA66FA9024CAD97FED6C7', query was: INSERT INTO `catalog_product_entity_tier_price` (`all_groups`, `customer_group_id`, `qty`, `value`, `website_id`, `percentage_value`, `row_id`) VALUES (?, ?, ?, ?, ?, ?, ?)"

---
Please provide [Severity](https://devdocs.magento.com/guides/v2.4/contributor-guide/contributing.html#backlog) assessment for the Issue as Reporter. This information will help during Confirmation and Issue triage processes.

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [x ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

**Comment by johan-lindahl** [comment]
The exception is thrown because getOrigData('tier_price') is returning tier_price data for the default store and not for the store edited when called by the following line.
https://github.com/magento/magento2/blob/96faea78235f05b5c74f78db9dc8d5661a38c2fd/app/code/Magento/Catalog/Model/Product/Attribute/Backend/TierPrice/UpdateHandler.php#L99

The reson why this happens is because the scope is not considered by the function initializeProductData on the following line

https://github.com/magento/magento2/blob/96faea78235f05b5c74f78db9dc8d5661a38c2fd/app/code/Magento/Catalog/Model/ProductRepository.php#L390

I think the scope must be used when initializing the product data, the correct implementation should then be

`$product = $this->getById($productData['id'], false, $productData['store_id']);
`

When I make the suggested change and add the store_id parameter to the function call above, the exception is not thrown and the product save as intended.

I hope someone with more knowledge about any possible side effects/consequences of the above suggested change could comment on this issue and maybe create a PR.

**Comment by engcom-Echo** [comment]
Hello @johan-lindahl,

Thanks for contribution and collaboration.
We have tried to reproduce this issue on magento 2.4-develop branch and it is not able to reproduce.
Steps followed to reproduce issue:

1- Install the fresh magento instance.
2- Create multiple website(Stores->All Stores->create website)
3- Set the "price scope" = website (Stores->Configuration->Catalog->Catalog->Price->Catalog Price Scope)
4. Add the Product with tier price.

Please find the attached screen for your reference.


<img width="1440" alt="image" src="https://user-images.githubusercontent.com/51681435/181015067-a6e81f84-3fef-45f9-8703-9021a7022209.png">

Please check if i missed anything and kindly check on magento 2.4-develop branch and provide more information so that we can reproduce this issue.

Thanks

**Comment by johan-lindahl** [comment]
The issue is with tier prices, the attached screen does not show tier price. Tier prices are found under "advanced pricing"

//Johan

**Comment by engcom-Echo** [comment]
Hello @johan-lindahl ,

Thanks for contribution and collaboration.
We have tried to reproduce this issue on magento 2.4-develop branch and it is not able to reproduce.
Steps followed to reproduce issue:

1- Install the fresh magento instance.
2- Create multiple website(Stores->All Stores->create website)
3- Set the "price scope" = website (Stores->Configuration->Catalog->Catalog->Price->Catalog Price Scope)
4. Add the Product with tier price.

We have add product with tier price with multiple website.
Please find the attached screen for your reference.

<img width="1440" alt="image" src="https://user-images.githubusercontent.com/51681435/181548241-94c2e37a-0ec4-47fe-acbe-dfd5cfda6acf.png">

<img width="1440" alt="image" src="https://user-images.githubusercontent.com/51681435/181549368-d7f3babd-4cb3-4fa1-88f2-adf35b20aeec.png">

Please check if i missed anything and kindly check on magento 2.4-develop branch and provide more information with exact steps and screenshot  so that we can reproduce this issue.

Thanks

**Comment by johan-lindahl** [comment]
You should try to change scope on the product page in the scope drop down (top left) and then set different tier price for different scopes.

//Johan

Sent from my iPhone

**Comment by engcom-Echo** [comment]
Hello @johan-lindahl ,

Thanks for contribution and collaboration.
We have tried to reproduce this issue on magento 2.4-develop branch and it is not able to reproduce.
Steps followed to reproduce issue:

1- Install the fresh magento instance.
2- Create multiple website(Stores->All Stores->create website)
3- Set the "price scope" = website (Stores->Configuration->Catalog->Catalog->Price->Catalog Price Scope)
4. Add the Product with tier price.

We have add product with tier price with multiple website.
Also, we have tried with change the scope and set the different tier price for different scopes for same product.
Please find the attached screen for your reference.

<img width="1440" alt="image" src="https://user-images.githubusercontent.com/51681435/182182919-608816c7-e37a-464c-9251-776440be066b.png">

<img width="1440" alt="image" src="https://user-images.githubusercontent.com/51681435/182183068-0dbebcc9-4fd6-46cf-9425-8f2c556adb24.png">

<img width="1440" alt="image" src="https://user-images.githubusercontent.com/51681435/182183168-3841f0ff-2c41-428e-a2f6-6e47754e0576.png">

<img width="1440" alt="image" src="https://user-images.githubusercontent.com/51681435/182183288-8576d058-7cab-45aa-a02a-8700f7e5bcdd.png">

<img width="1440" alt="image" src="https://user-images.githubusercontent.com/51681435/182183398-151066a5-f103-4a3e-9dba-feb0cee75d89.png">

<img width="1440" alt="image" src="https://user-images.githubusercontent.com/51681435/182183652-448b1f48-3f2e-4210-919c-5293b79f0b26.png">


Please check if i missed anything and kindly check on magento 2.4-develop branch and provide more information with exact steps and screenshot or video (if possible) so that we can try to reproduce this issue at our end.

Thanks

**Comment by engcom-Echo** [comment]
Hello @johan-lindahl,

We have noticed that this issue has not been updated for a period of 14 Days.
Hence we assume that this issue is fixed now, so we are closing it.Please raise a fresh ticket or reopen this ticket if you need more assistance on this.

Thanks

**Comment by ereztaiar** [comment]
Had the same issue on version 2.4.5-p1 when updating product via code
Looks like the admin controller to save product uses the deprecated function save on product which does not create a new product object when saving
But when you save the product with a product repository then a new product object is created that new product does not have the attribute tier_prices in it and so the data is being inserted instead of being updated and fail on duplicate data

I had to change my code to use product->save() so it wont fail on products with tier prices

**Comment by benyatesayko** [comment]
Seeing this issue on 2.4.6-p6 when updating via the rest API.

A product that already has tiered prices saved against the global scope throws this error, even when we're not updating the tier prices in the PUT update request.

As others have observed, the below line in wrongly returns no results vendor/magento/module-catalog/Model/Product/Attribute/Backend/TierPrice/UpdateHandler.php
`            $origPrices = $entity->getOrigData($attribute->getName());
`

**Comment by benyatesayko** [comment]
In my specific case, it looks like the issue was with type_id. In my case, there were products with a type ID of 'Simple' instead of 'simple'.

Manually updating these in the database seemed to allow us to update via the rest API afterwards. I think this type ID must have been messed up after having initially been created with tier prices, from there it doesn't seem to be possible to rectify this via the rest API.

**Comment by GautierDig** [comment]
Same issue here ! Any news on this ?

**Comment by benyatesvortex** [comment]
I'm seeing this on another project now. It's showing some bizarre behaviour. I have the product repository within a loop and is loading by sku (i.e. the get method). It seemed to fail on the same products each time, making me think it was data related, but what I'm finding is if I change the order in which the products are loaded, then different product will succeed/fail.

**Comment by benyatesvortex** [comment]
No resolution on this yet, but I've isolated the issue from my code entirely and tried running this in the magerun2 dev console, and I'm seeing the same issue:

```
/* We initialise our classes using objectmanager (because we're in the dev console)
$state = \Magento\Framework\App\ObjectManager::getInstance()->get( \Magento\Framework\App\State::class);
$state->setAreaCode('adminhtml');
$productRepository = \Magento\Framework\App\ObjectManager::getInstance()->get(\Magento\Catalog\Model\ProductRepository::class);
   
   
/* We retrieve the first product and save, this works fine */   
$product = $productRepository->get('311512');
$productRepository->save($product);

/* We retrieve the second product and save, we get an integrity constraint violation*/
$product = $productRepository->get('311448');
$productRepository->save($product);
```
My assumption would be that there's a data issue with the second product, as was the case the first time I've had this issue. But if I start the dev console again and try load/save the products in the reverse order, the first product 311448 succeeds and 311512 fails. So the products after the first load seem to have issues. When in the dev console I also found that if I tried to save the second product a second time, it would generally succeed.

**Comment by bavaba** [comment]
Hi all, this patch helped me to solve the issue on the 2.4.6-p6 version 

https://github.com/magento/quality-patches/blob/master/patches/os/ACSD-51845_2.4.5-p2.patch

Hope it will help

**Comment by benyatesvortex** [comment]
Thanks @bavaba ! That's fixed it for me.

 I think the description in https://experienceleague.adobe.com/en/docs/commerce-knowledge-base/kb/support-tools/patches/v1-1-35/acsd-51845-cant-update-subsequent-products-with-tier-prices-and-different-attribute-sets-via-asynch-bulk-api could probably do with updating as it's not necessarily anything to do with the Bulk API. But good to hear this is fixed from 2.4.7 onwards.

**Comment by Charbel-magebit** [comment]
Any progress on this? The patch mentioned [here](https://github.com/magento/magento2/issues/35444#issuecomment-2346075589) does not work anymore for newer versions, it seems.

I'm on  2.4.6-p12, but it seems like the quickest and least intrusive solution is to use the deprecated $product->save().

The fix suggested [here](https://github.com/magento/magento2/issues/35444#issuecomment-2966116082) seems reasonable but I'm not sure if it may impact other places that maybe were built on this bug being there
